src/lib/parser.c

```diff
@@ -750,40 +759,118 @@ static inline RdbStatus unpackList(RdbParser *p, unsigned char *lp) {
+         *    - Supply the referenced, or copied, application bulk to callback
+         *    - And finalize by restoring original char from `tmp`.
+         */
+
+        itemEnd = item + stringLen;
+        tmp = *itemEnd;
+        IF_NOT_OK_RETURN(allocFromCache(p, stringLen, RQ_ALLOC_APP_BULK_REF, (char *) item, &binfo));
 
-    if (!ziplistGet(p, &str, &slen, &vll)) return 0;
+    } else {
+        int buflen = 32;
+        IF_NOT_OK_RETURN(allocFromCache(p, buflen, RQ_ALLOC_APP_BULK, NULL, &binfo));
+        binfo->len  = ll2string(binfo->ref, buflen, itemVal);  /* update len */
 
-    if (str)
-        *lp = lpAppend(*lp, (unsigned char*)str, slen);
-    else
-        *lp = lpAppendInteger(*lp, vll);
+        /* set itemEnd to point a dummy char. CALL_HANDLERS_CB goanna write it on finalize */
+        itemEnd = &dummy;
+    }
 
-    return 1;
+    registerAppBulkForNextCb(p, binfo);
+    CALL_HANDLERS_CB(p,
+                     *itemEnd = tmp,     /* <<< finalize: restore modified char */
+                     RDB_LEVEL_DATA,
+                     rdbData.handleListElement,
+                     binfo->ref);
+    return 0;
+}
+
+/* in order to distinct between eof list and an error, the caller to integrity
+ * check with this func need to check on return if (p->errorCode != 0) */
+static int ziplistItemCallback(unsigned char *ptr, unsigned int head_count, void *userdata) {
+    return (0 != ziplistItemCallbackInner(ptr, head_count, userdata)) ? 0 : 1;
 }
 
-/* return either OK or ERROR */
-static RdbStatus listpackCallback(RdbParser *p, BulkInfo *lpInfo) {
+/* return either RDB_STATUS_OK or RDB_STATUS_ERROR */
+static RdbStatus listListpackItem(RdbParser *p, BulkInfo *lpInfo) {
     /* Silently skip empty listpack */
     if (lpLength(lpInfo->ref) == 0) return RDB_STATUS_OK;
 
     if (p->elmCtx.key.handleByLevel == RDB_LEVEL_STRUCT) {
         registerAppBulkForNextCb(p, lpInfo);
-        CALL_HANDLERS_CB(p, NOP, RDB_LEVEL_STRUCT, rdbStruct.handleQListNode, lpInfo->ref);
+        CALL_HANDLERS_CB(p, NOP, RDB_LEVEL_STRUCT, rdbStruct.handleListLP, lpInfo->ref);
     } else {
         /* unpackList makes multiple callbacks. all data in ctx.lp */
         IF_NOT_OK_RETURN(unpackList(p, lpInfo->ref));
     }
     return RDB_STATUS_OK;
 }
 
+/* return either RDB_STATUS_OK or RDB_STATUS_ERROR */
+static RdbStatus listZipListItem(RdbParser *p, BulkInfo *ziplistBulk) {
+    int ret;
+
+    if (p->elmCtx.key.handleByLevel == RDB_LEVEL_STRUCT) {
+        ret = ziplistValidateIntegrity(ziplistBulk->ref, ziplistBulk->len, p->deepIntegCheck, NULL, NULL);
+
+        if (unlikely(!ret)) {
+            RDB_reportError(p, RDB_ERR_ZIP_LIST_INTEG_CHECK, "elementZipList(): ZipList integrity check failed");
+            return RDB_STATUS_ERROR;
+        }
+
+        registerAppBulkForNextCb(p, ziplistBulk);
+        CALL_HANDLERS_CB(p, NOP, RDB_LEVEL_STRUCT, rdbStruct.handleListLP, ziplistBulk->ref);
```

Comment:
is this a bug?
we're inside `listZipListItem`, and use `handleListLP`
p.s. i think the "L" in ZipList should be lower case, i.e. `ziplist` or `Ziplist`.
or maybe it's a good idea to change them all to `ZL` and `LP`, and refrain from using the full "ziplist" or "listpack" in API or variable names?

Reply:
it is a bug. 

The reason that the test passes is because the json callbacks makes trivial write of the data, without caring much about what is inside the payload.